include/envoy/network/transport_socket.h

```diff
@@ -0,0 +1,118 @@
+#pragma once
+
+#include "envoy/buffer/buffer.h"
+#include "envoy/common/pure.h"
+#include "envoy/network/connection.h"
+
+namespace Envoy {
+namespace Network {
+
+/**
+ * Action that should occur on a connection after I/O.
+ */
+enum class PostIoAction {
+  // Close the connection.
+  Close,
+  // Keep the connection open.
+  KeepOpen
+};
+
+/**
+ * Result of each I/O event.
+ */
+struct IoResult {
+  PostIoAction action_;
+
+  /**
+   * Number of bytes processed by the I/O event.
+   */
+  uint64_t bytes_processed_;
+};
+
+/**
+ * Callbacks used by transport socket instances to communicate with connection
+ */
+class TransportSocketCallbacks {
+public:
+  virtual ~TransportSocketCallbacks() {}
+
+  /**
+   * @return the file descriptor associated with the connection.
+   */
+  virtual int fd() PURE;
+
+  /**
+   * @return the connection interface.
+   */
+  virtual Connection& connection() PURE;
+
+  /**
+   * @return return whether the read buffer should be drained
+   */
+  virtual bool shouldDrainReadBuffer() PURE;
+
+  /**
+   * Mark read buffer ready to read in the event loop. This is used when yielding following
+   * shouldDrainReadBuffer().
+   */
+  virtual void setReadBufferReady() PURE;
+
+  /**
+   * Raise a connection event to the connection.
+   * @param event supplies the connection event
+   */
+  virtual void raiseEvent(ConnectionEvent event) PURE;
```

Comment:
If this is done on the connection, I'd lean towards exposing it there rather than having a wrapper function here.  I guess since this is not accessible to all entities that have a connection it could be useful to scope it?  Your call.

Either way you could doc up a bit more what this does.  It basically fakes the connection getting the specific event from the kernel, no?

Reply:
Done